src/Bounds.cpp

```diff
@@ -1174,7 +1196,88 @@ class BoxesTouched : public IRGraphVisitor {
     }
 
     void visit(const LetStmt *op) {
+        CollectVars collect(op->name);
+        op->value.accept(&collect);
+        for (const auto &v : collect.vars) {
+            children[v].insert(op->name);
+        }
+        let_stmts.push(op->name, op->value);
+
         visit_let(op);
+
+        let_stmts.pop(op->name);
+        for (const auto &v : collect.vars) {
+            children[v].erase(op->name);
+        }
+    }
+
+    struct LetBound {
+        string var, min_name, max_name;
+        Expr min_value, max_value;
+        LetBound(const string &v, const string &min, const string &max,
+                 const Expr &min_val, const Expr &max_val)
+            : var(v), min_name(min), max_name(max), min_value(min_val), max_value(max_val) {}
+    };
+
+    vector<LetBound> trim_scope_push(const string &name, const Interval &bound) {
+        scope.push(name, bound);
+
+        vector<LetBound> let_bounds;
+        for (const auto &v : children[name]) {
+            internal_assert(scope.contains(v));
+            internal_assert(let_stmts.contains(v));
+
+            const Expr &val = let_stmts.get(v);
+            Interval v_bound = bounds_of_expr_in_scope(val, scope, func_bounds);
+            bool fixed = v_bound.min.same_as(v_bound.max);
+            v_bound.min = simplify(v_bound.min);
+            v_bound.max = fixed ? v_bound.min : simplify(v_bound.max);
+
+            string max_name = unique_name('t');
+            string min_name = unique_name('t');
+
+            internal_assert(scope.contains(v));
+            const Interval &old_bound = scope.get(v);
+            Expr max_val = simplify(min(v_bound.max, old_bound.max));
+            Expr min_val = simplify(max(v_bound.min, old_bound.min));
+
+            let_bounds.push_back(LetBound(v, min_name, max_name, min_val, max_val));
+
+            scope.push(v, Interval(Variable::make(val.type(), min_name),
```

Comment:
Should this line be trim_scope_push instead of scope.push? 

I'm thinking of whether or not z gets trimmed in the case below:

let x = y+2
let z = x+2
if (y < 4) {
 ...
}